### ISSUE TITLE
New Rule: prefer-return-typedef

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint#1803]
### Rule suggestion

I sometimes see unnecessary typedefs or type assertions in return statements that instead belong on the method signature directly:

```ts
// bad
function makeBlockScope() {
    return {
        ...
    } as IBlockScope;
}

// good
function makeBlockScope(): IBlockScope {
    return {
        ...
    };
}
```

**Comment by WilcoBakker** [comment]
+1

**Comment by dploeger** [comment]
@adidahiya This is somewhat what I was searching. My rule suggestion would be a rule, that forces a return type definition on each function/method. Can you also agree to this? 

Or do you specifically want to alert on returns with typedefs without matching typedefs on the function/method?

My good/bad example would be this:

```typescript

// bad
function something() {
  console.log('bla')
}

// good
function something(): void {
  console.log('bla')
}
```

My use case is, that the developer tends to forget to write the function typedef.

**Comment by dploeger** [comment]
@adidahiya ARGH. I missed that and was looking for it for a long time. Thanks, mate!

**Comment by JoshuaKGoldberg** [comment]
Note: per #4534, this issue will be closed in less than a month if no PR is sent to add the rule. If you _really_ need the rule, custom rules are always an option and can be maintained outside this repo!

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated** and no longer accepting pull requests for new rules. See #4534. 😱

If you'd like to see this rule implemented, you have two choices:
* **Recommended:** Use [typescript-eslint](https://typescript-eslint.io) to lint your TypeScript code with ESLint ✅
* **Legacy:** Write it as a [custom TSLint rule](https://palantir.github.io/tslint/develop/custom-rules/) 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
